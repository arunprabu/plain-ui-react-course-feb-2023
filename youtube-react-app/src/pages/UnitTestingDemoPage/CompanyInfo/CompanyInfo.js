import PropTypes from 'prop-types';
import React, { useState } from 'react';

const CompanyInfo = ({ foundedYear }) => {
  const [countryName, setCountryName] = useState('USA');

  return (
    <div>
      <h2>Company Info! | Testing Props, Styles, Placeholder, Events, States and Snapshot</h2>
      <h3 data-testid='companyName'>Company Name: Cognizant</h3>
      <p
        data-testid='foundedYear'
        style={{
          color: 'rgb(0, 255, 0)',
          backgroundColor: 'black'
        }}>
        Founded in: {foundedYear}
      </p>
      <button className='btn btn-primary'>JOIN COGNIZANT</button>
      <hr />

      <input
        type='text'
        placeholder='Enter Your Country Name!'
        value={countryName}
        onChange={(event) => {
          setCountryName(event.target.value);
        }}
      />
      <p data-testid='visitWebsite'>Please visit Cognizant {countryName} Website</p>
    </div>
  );
};

CompanyInfo.propTypes = {
  foundedYear: PropTypes.string
};

export default CompanyInfo;
