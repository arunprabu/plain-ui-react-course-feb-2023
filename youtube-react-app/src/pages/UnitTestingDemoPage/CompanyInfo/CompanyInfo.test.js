import { render, screen, fireEvent } from '@testing-library/react';
import CompanyInfo from './CompanyInfo';
import renderer from 'react-test-renderer';

describe('CompanyInfo', () => {
  // test spec
  it('has Cognizant as company name', () => {
    // act
    render(<CompanyInfo />);
    const companyName = screen.getByTestId('companyName');
    console.log(companyName.textContent);
    // assert
    expect(companyName.textContent).toBe('Company Name: Cognizant');
  });

  // testing props -- must
  it('receives foundedYear prop and displays in JSX', () => {
    render(<CompanyInfo foundedYear='2012'/>);
    const foundedYear = screen.getByTestId('foundedYear');
    expect(foundedYear).toHaveTextContent('Founded in: 2012');
  });

  // testing inline styles
  it('has foundedYear with text color green -- inline styles', () => {
    render(<CompanyInfo foundedYear='2012' />);
    const foundedYear = screen.getByTestId('foundedYear');
    // recommended to use color code in hexa or rgb -- not the color name
    expect(foundedYear).toHaveStyle('color: #00ff00');
  });

  // TODO: testing whether a button has css class btn-primary
  // the button should have text -- JOIN COGNIZANT
  // use data-testid
  // you can also have multiple expectations

  // let's find out whether the input has the right placeholder or not.

  it('has an input with `Enter Your Country Name` as placeholder', () => {
    render(<CompanyInfo />);
    const countryInput = screen.getByPlaceholderText('Enter Your Country Name!');
    expect(countryInput).toBeTruthy();
  });

  it('displays countryName USA by default and updates JSX onChange of country input', () => {
    render(<CompanyInfo />);
    // finding input element
    const countryInput = screen.getByPlaceholderText('Enter Your Country Name!');
    console.log(countryInput.value);
    expect(countryInput.value).toBe('USA');

    // now checking whether the onChange event handler is working or not
    // setting up mock event obj
    const mockEventObj = {
      target: {
        value: 'UK'
      }
    }

    // as the comp is rendered in test runner we can't use mouse or keyboard as input devices.
    // so, let's trigger the event thru program
    fireEvent.change(countryInput, mockEventObj);
    // finally let's check if the input element is showing the change
    expect(countryInput.value).toBe('UK');

    expect(screen.getByTestId('visitWebsite')).toHaveTextContent(
      'Please visit Cognizant UK Website'
    );
  });

  // snapshot testing
  it('has right snapshot with all requirements completed', () => {
    // to take snapshot we need a react-test-renderer  // npm i react-test-renderer -D
    // taking snapshot and converting it into JSON
    // [RECOMMENDED]: Take snapshot with all possible props as well as props children
    const snapshotInJson = renderer.create(<CompanyInfo foundedYear='2006' />).toJSON();
    // let's assert with toMatchSnapshot
    expect(snapshotInJson).toMatchSnapshot();
  });
});
