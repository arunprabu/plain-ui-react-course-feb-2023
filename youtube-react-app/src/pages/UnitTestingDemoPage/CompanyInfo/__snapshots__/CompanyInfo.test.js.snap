// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyInfo has right snapshot with all requirements completed 1`] = `
<div>
  <h2>
    Company Info! | Testing Props, Styles, Placeholder, Events, States and Snapshot
  </h2>
  <h3
    data-testid="companyName"
  >
    Company Name: Cognizant
  </h3>
  <p
    data-testid="foundedYear"
    style={
      Object {
        "backgroundColor": "black",
        "color": "rgb(0, 255, 0)",
      }
    }
  >
    Founded in: 
    2006
  </p>
  <button
    className="btn btn-primary"
  >
    JOIN COGNIZANT
  </button>
  <hr />
  <input
    onChange={[Function]}
    placeholder="Enter Your Country Name!"
    type="text"
    value="USA"
  />
  <p
    data-testid="visitWebsite"
  >
    Please visit Cognizant 
    USA
     Website
  </p>
</div>
`;
